docker/registry/dockerhub.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package registry
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/docker"
+)
+
+const (
+	dockerServerAddress = "index.docker.io"
+)
+
+type dockerhub struct {
+	*baseClient
+}
+
+func newDockerhub(repoDetails docker.ImageRepoDetails, transport http.RoundTripper) RegistryInternal {
+	c := newBase(repoDetails, DefaultTransport)
+	return &dockerhub{c}
+}
+
+func (c *dockerhub) Match() bool {
+	c.prepare()
+	return c.repoDetails.ServerAddress == "" || strings.Contains(c.repoDetails.ServerAddress, "docker.io")
+}
+
+func (c *dockerhub) DecideBaseURL() error {
```

Comment:
`DecideBaseURL` seems like it should be called something like `Initialise` or something. 
Actually, the functionality here seems to pertain to the `repoDetails` value - should this be an internal method used to set up things whenever the `repoDetails` is set, ie should this be called from `newDockerHub` and not exposed as a separate method?

Reply:
DecideBaseURL is exported for generating mocks and it's one of the methods called in `initClient`.
So DecideBaseURL is just one of the steps of Initialise.
I will move the provider integration code to `internal` sub-package in the following PR.